ccw.core/src/clj/ccw/leiningen/classpath_container.clj

```diff
@@ -353,50 +364,102 @@
-            ;; it can take a long time, so we do not put it inside the workspace job which blocks on the workspace root
-            (doto
-              (e/workspace-job
-                (format "Upgrade project build path for project %s" (e/project-name java-project))
-                (fn [^IProgressMonitor monitor]
-                  (set-lein-container java-project deps)
-                  (delete-container-markers java-project)
-                  (save-project-dependencies java-project deps)
-                  (doto ;; we refresh the target folder outside the workspace-root lock
-                    (e/workspace-job
-                      (format "Refreshing project %s" (e/project-name java-project))
-                      (fn [^IProgressMonitor monitor]
-                        (refresh-target-folder java-project monitor)))
-                    (.setUser true)
-                    (.schedule))))
-              (.setUser true)
-              ;; this rule is OK because we know the job needs it and will not take too long
-              (.setRule (e/workspace-root))
-              (.schedule)))
-          (catch Exception e
-            ;; TODO enhance this in the future ... (more accurate problem markers)
-            (let [[jresource message] (resource-message e java-project)
-                  project-name (-> java-project e/resource .getName)]
-              (report-container-error
-                jresource
-                (format "Leiningen Managed Dependencies issue: %s" message)
-                e))))))
-    (.setUser true)
-    (.schedule)))
+  (let [project-id (e/identifier java-project)
+        project-name (e/project-name java-project)
+        error-fn (fn [[workspace-job exception]]
+                   (let [[jresource message] (resource-message exception java-project)
+                         error-message (format "Leiningen Managed Dependencies issue: %s" message)]
+                     ;; TODO enhance this in the future ... (more accurate problem markers)
+                     (report-container-error jresource error-message exception)
+                     error-message))
+        finally-fn (fn [workspace-job] (doto ;; we refresh the target folder outside the workspace-root lock
+                             (e/workspace-job
+                              (format "Refreshing project %s" project-name)
+                              (fn [^IProgressMonitor monitor]
+                                (refresh-target-folder java-project monitor)
+                                Status/OK_STATUS))
+                           (.setUser true)
+                           (.schedule)))
+        job (e/cancelable-workspace-job
+             (format "Update project dependencies for project %s" project-name)
+             (fn [^IProgressMonitor monitor]
+               (letfn [(cancel! [] (e/throw-op-canceled! (format "The update of %s was canceled" project-name) monitor))]
+                 (let [_ (cancel!)
```

Comment:
Basically it should cancel the execution when appropriate.
Now it throws when monitor's ```isCanceled``` is true.